### PR TITLE
[EOC-31] Create new shopping list

### PR DESCRIPTION
**Pull request by adamklepacz**
Create new shopping list and store it in the db.
zrzut ekranu 2019-01-25 o 11 53 51

### CODE REVIEW

**Comment by adamklepacz** [comment]
@marekrozmus  You're right I have moved code to shopping-list module

File: client/src/app/model/reducer.js [hunk 1]
```diff
@@ -1,8 +1,11 @@
 import { combineReducers } from 'redux';
 
-import items, { uiStatus } from 'modules/shopping-list/model/reducer';
+import items, {
```
Comment (alemikolo) [inline client/src/app/model/reducer.js]: shouldn't we have two separate reducers for items (products) and shopping lists?

File: client/src/modules/shopping-list/model/reducer.js [hunk 2]
```diff
@@ -33,7 +38,18 @@ const items = (state = itemsInitialState, action) => {
   }
 };
 
-const uiStatus = (state = initialStatus, action) => {
+export const shoppingLists = (state = [], action) => {
```
Comment (alemikolo) [inline client/src/modules/shopping-list/model/reducer.js]: shouldn't we have two separate reducers for items (products) and shopping lists?

File: server/controllers/shoppingList.js [hunk 3]
```diff
@@ -20,11 +20,16 @@ const createNewList = (req, resp) => {
...
-      ? resp.status(401).send(err.message)
-      : resp.status(200).send(shoppingLists);
-  });
+  ShoppingList.find(
+    {},
+    null,
+    { sort: { created_at: -1 } },
+    (err, shoppingLists) => {
+      err
+        ? resp.status(401).send(err.message)
```
Comment (alemikolo) [inline server/controllers/shoppingList.js]: status 401 - unauthorized, I'm not sure if controller check authorization...

File: client/src/modules/shopping-list/model/reducer.js (hunk 2 above)
Comment (adamklepacz) [reply client/src/modules/shopping-list/model/reducer.js]: yes we have separated reducers for items/products and shopping lists

Comment (alemikolo) [reply client/src/modules/shopping-list/model/reducer.js]: Sorry, I meant that in in separate files

File: client/src/modules/shopping-list/model/actions.js [hunk 4]
```diff
@@ -3,14 +3,31 @@ import { ENDPOINT_URL } from 'common/constants/variables';
...
+});
 
 // Dispatchers
 export const fetchItems = () => dispatch =>
   fetch(`${ENDPOINT_URL}/items`, { credentials: 'same-origin' })
     .then(resp => resp.json())
     .then(json => dispatch(recieveItems(json)))
     .catch(err => dispatch(fetchItemsError(err)));
+
+export const setNewShoppingList = (name, description) => dispatch =>
```
Comment (marekrozmus) [inline client/src/modules/shopping-list/model/actions.js]: again - what does it `set`?

File: client/src/modules/shopping-list/model/actions.js [hunk 5]
```diff
@@ -3,14 +3,31 @@ import { ENDPOINT_URL } from 'common/constants/variables';
 // Action types
 export const FETCH_FAILED = 'FETCH_FAILED';
 export const FETCH_ITEMS = 'FETCH_ITEMS';
+export const ADD_SHOPPING_LIST_SUCCESS = 'ADD_NEW_SHOPPING_LIST';
 
 // Action creators
 export const fetchItemsError = err => ({ type: FETCH_FAILED, err });
 export const recieveItems = json => ({ type: FETCH_ITEMS, items: json });
+const newShoppingListSuccess = data => ({
```
Comment (marekrozmus) [inline client/src/modules/shopping-list/model/actions.js]: `new` what? there is no verb in this action name

File: client/src/modules/shopping-list/model/actions.js [hunk 6]
```diff
@@ -3,14 +3,31 @@ import { ENDPOINT_URL } from 'common/constants/variables';
...
 
 // Dispatchers
 export const fetchItems = () => dispatch =>
   fetch(`${ENDPOINT_URL}/items`, { credentials: 'same-origin' })
     .then(resp => resp.json())
     .then(json => dispatch(recieveItems(json)))
     .catch(err => dispatch(fetchItemsError(err)));
+
+export const setNewShoppingList = (name, description) => dispatch =>
+  fetch(`${ENDPOINT_URL}/shopping-lists/new-list`, {
```
Comment (marekrozmus) [inline client/src/modules/shopping-list/model/actions.js]: consider create some helper `post` method - you will use it in many places and only payload and endpoint will change

File: client/src/modules/shopping-list/model/reducer.js [hunk 7]
```diff
@@ -1,5 +1,9 @@
 import { items as itemsInitialState, initialStatus } from './initialState';
-import { FETCH_FAILED, FETCH_ITEMS } from 'modules/shopping-list/model/actions';
+import {
+  FETCH_FAILED,
+  FETCH_ITEMS,
+  ADD_SHOPPING_LIST_SUCCESS
+} from './actions';
```
Comment (marekrozmus) [inline client/src/modules/shopping-list/model/reducer.js]: action should be enum and imported as `ShoppingListActionTypes`. Currently in code we have no idea what action `FETCH_FAILED' really mean.

File: client/src/modules/dashboard/index.jsx [hunk 8]
```diff
@@ -6,15 +6,18 @@ import {
   ClipboardSolid as ShoppingListIcon,
   UsersSolid as CohortIcon
 } from 'assets/images/icons';
-import { fetchAllLists } from './model/actions';
-import { getCohorts, getShoppingLists } from './model/selectors';
+import { fetchShoppingLists } from 'modules/shopping-list/model/actions';
+import { fetchAllCohorts } from 'modules/cohort/model/actions';
```
Comment (alemikolo) [inline client/src/modules/dashboard/index.jsx]: fetchShoppingLists (1) but fetchAllCohorts (2) maybe add (1) or remove (2) 'All' - that the names would be more consistent?

File: client/src/modules/shopping-list/components/InputBar/enum.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export const ItemActionTypes = Object.freeze({
+  ADD_ITEM: 'item/ADD_ITEM',
+  ADD_ITEM_FAILURE: 'item/ADD_ITEM_FAILURE',
+  ADD_ITEM_SUCCESS: 'item/ADD_ITEM_SUCCESS',
+  TOGGLE_ITEM: 'item/TOGGLE_ITEM',
+  OTE_FOR_ITEM: 'item/VOTE_FOR_ITEM'
```
Comment (alemikolo) [inline client/src/modules/shopping-list/components/InputBar/enum.js]: VOTE_FOR_ITEM

File: client/src/modules/shopping-list/components/InputBar/model/actions.js [hunk 10]
```diff
@@ -1,14 +1,16 @@
...
-export const ADD_ITEM = 'ADD_ITEM';
-export const ADD_ITEM_FAILURE = 'ADD_ITEM_FAILURE';
-export const ADD_ITEM_SUCCESS = 'ADD_ITEM_SUCCESS';
+import { ItemActionTypes } from '../enum';
 
 // Action creators
-export const addItemError = err => ({ type: ADD_ITEM_FAILURE, err });
-export const addItemSuccess = item => ({ type: ADD_ITEM_SUCCESS, item });
+export const addItemError = err => ({
+  type: ItemActionTypes.ADD_ITEM_FAILURE,
```
Comment (alemikolo) [inline client/src/modules/shopping-list/components/InputBar/model/actions.js]: addItemSuccess -> ADD_ITEM_SUCCESS
addItemError -> ADD_ITEM_FAILURE

Comment (adamklepacz) [reply client/src/modules/shopping-list/components/InputBar/model/actions.js]: What you mean?

File: client/src/modules/cohort/model/reducer.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { CohortActionTypes } from '../enum';
+import { initalCohorts } from './initialState';
+
+export const cohortsList = (state = initalCohorts, action) => {
```
Comment (marekrozmus) [inline client/src/modules/cohort/model/reducer.js]: leave just `cohorts`

File: client/src/modules/cohort/model/reducer.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  switch (action.type) {
+    case CohortActionTypes.FETCH_ALL_COHORTS_SUCCESS:
```
Comment (marekrozmus) [inline client/src/modules/cohort/model/reducer.js]: remove `_ALL_`

File: client/src/common/utils/fetchMethods.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
+export const getData = url =>
+  fetch(url, {
+    credentials: 'same-origin'
+  });
+
+export const postData = (url, data) =>
+  fetch(url, {
+    body: data,
```
Comment (marekrozmus) [inline client/src/common/utils/fetchMethods.js]: add JSON.stringify here

File: client/src/common/utils/fetchMethods.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    method: 'POST'
+  });
+
+export const patchData = (url, data) =>
+  fetch(url, {
+    body: data,
```
Comment (marekrozmus) [inline client/src/common/utils/fetchMethods.js]: add JSON.stringify here

**Pull request merged**